blog/en/blog/2023/03/09/authenticate-openid-connect.md

```diff
@@ -0,0 +1,195 @@
+>
+>-- [What is OpenID Connect?](https://openid.net/connect/)
+
+Here are a couple of identity providers that are compatible with OpenID Connect:
+
+* GitHub
+* Google
+* Microsoft
+* Apple
+* Facebook
+* Twitter
+* Spotify
+
+In the following, we will start with Google and switch to Azure to validate our setup.
+
+## Setting up OpenID Connect with Apache APISIX
+
+Imagine we have a web app behind Apache APISIX that we want to secure with OpenID Connect. Here's the corresponding Docker Compose file:
+
+```yaml
+version: "3"
+
+services:
+  apisix:
+    image: apache/apisix:3.1.0-debian                              #1
+    ports:
+      - "9080:9080"
+    volumes:
+      - ./apisix/config.yml:/usr/local/apisix/conf/config.yaml:ro  #2
+      - ./apisix/apisix.yml:/usr/local/apisix/conf/apisix.yaml:ro  #3
+    env_file:
+      - .env
+  httpbin:
+    image: kennethreitz/httpbin                                    #4
+```
+
+1. Apache APISIX API Gateway
+2. APISIX configuration - used to configure it statically in the following line
+3. Configure the single route
+4. Webapp to protect. Any will do
+
+Apache APISIX offers a plugin-based architecture. One such plugin is the [openid-connect](https://apisix.apache.org/docs/apisix/plugins/openid-connect/) plugin, which allows using OpenID Connect.
+
+Let's configure it:
+
+```yaml
+routes:
+  - uri: /*                                                                    #1
+    upstream:
+      nodes:
+        "httpbin:80": 1                                                        #1
+    plugins:
+      openid-connect:
+        client_id: ${{OIDC_CLIENTID}}                                          #2
+        client_secret: ${{OIDC_SECRET}}                                        #2
+        discovery: https://${{OIDC_ISSUER}}/.well-known/openid-configuration   #2-3
+        redirect_uri: http://localhost:9080/callback                           #4
+        scope: openid                                                          #5
+        session:
+          secret: ${{SESSION_SECRET}}                                          #6
+#END
+```
+
+1. Catch-all route to the underlying web app
+2. Plugin configuration parameters. Values depend on the exact provider (see below)
+3. OpenID Connect can use a Discovery endpoint to get all necessary OAuth endpoints. See [OpenID Connect Discovery 1.0 spec](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig) for more information
+4. Where to redirect when the authentication is successful. It mustn't clash with any of the explicitly defined routes. The plugin creates a dedicated route there to work its magic.
+5. Default scope
+6. Key to encrypt session data. Put whatever you want.
```

Comment:
When the secret is too short, the following error log will appear. Do we need to specify here that the key length is at least 16?
```
err:failed to check the configuration of plugin openid-connect err: property "session" validation failed: property "secret" validation failed: string too short, expected at least 16, got 4 
```